cni/cmd/tc-redirect-tap/main_test.go

```diff
@@ -0,0 +1,296 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"testing"
+
+	"net"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/containernetworking/cni/pkg/version"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/require"
+	"github.com/vishvananda/netlink"
+
+	"github.com/firecracker-microvm/firecracker-go-sdk/cni/internal"
+)
+
+func TestAdd(t *testing.T) {
+	vmID := "this-is-not-a-machine"
+
+	redirectInterfaceName := "veth0"
+	redirectMTU := 1337
+
+	redirectMac, err := net.ParseMAC("22:33:44:55:66:77")
+	require.NoError(t, err, "failed to get redirect mac")
+
+	tapName := "tap0"
+	tapMac, err := net.ParseMAC("11:22:33:44:55:66")
+	require.NoError(t, err, "failed to get tap mac")
+
+	tapUID := 123
+	tapGID := 456
+
+	netNS := internal.MockNetNS{MockPath: "/my/lil/netns"}
+
+	p := &plugin{
+		NetlinkOps: &internal.MockNetlinkOps{
+			CreatedTap: &internal.MockLink{
+				LinkAttrs: netlink.LinkAttrs{
+					Name:         tapName,
+					HardwareAddr: tapMac,
+				},
+			},
+			RedirectIface: &internal.MockLink{
+				LinkAttrs: netlink.LinkAttrs{
+					Name:         redirectInterfaceName,
+					HardwareAddr: redirectMac,
+					MTU:          redirectMTU,
+				},
+			},
+		},
+		vmID:                  vmID,
+		tapName:               tapName,
+		tapUID:                tapUID,
+		tapGID:                tapGID,
+		redirectInterfaceName: redirectInterfaceName,
+		netNS:                 netNS,
+	}
+
+	redirectIfacesIndex := 0
+
+	baseResult := &current.Result{
+		CNIVersion: version.Current(),
+		Interfaces: []*current.Interface{{
+			Name:    "veth0",
+			Sandbox: netNS.Path(),
+			Mac:     redirectMac.String(),
+		}},
+		IPs: []*current.IPConfig{{
+			Version:   "4",
+			Interface: &redirectIfacesIndex,
+			Address: net.IPNet{
+				IP:   net.IPv4(10, 0, 0, 2),
+				Mask: net.IPv4Mask(255, 255, 255, 0),
+			},
+			Gateway: net.IPv4(10, 0, 0, 1),
+		}},
+	}
+
+	newResult := &current.Result{
+		CNIVersion: baseResult.CNIVersion,
+		Interfaces: append([]*current.Interface{}, baseResult.Interfaces...),
+		IPs:        append([]*current.IPConfig{}, baseResult.IPs...),
+	}
+
+	err = p.add(newResult)
+	require.NoError(t, err,
+		"failed to add tap device")
+
+	require.Len(t, newResult.Interfaces,
```

Comment:
For checking the value which you can access without any additional function calls, I'm inclined to use `assert` than `require`, since it would be easier to read.

I think assert/require (or Fail/FailNow on testing package) are coming from googletest, which explicitly recommends non-fatal versions over fatal-versions;

https://github.com/google/googletest/blob/master/googletest/docs/primer.md

> The assertions come in pairs that test the same thing but have different effects on the current function. ASSERT_* versions generate fatal failures when they fail, and abort the current function. EXPECT_* versions generate nonfatal failures, which don't abort the current function. Usually EXPECT_* are preferred, as they allow more than one failure to be reported in a test. However, you should use ASSERT_* if it doesn't make sense to continue when the assertion in question fails.

Reply:
Yeah, agree that's useful, updated all the test files